CIP-optimized-uplc-serialization/README.md

```diff
@@ -0,0 +1,327 @@
+---
+CIP: "???"
+Title: optimized UPLC serialization  
+Authors: Michele Nuzzi <michele.nuzzi.2014@gamil.com> 
+Discussions-To: TBD
+Comments-Summary: TBD  
+Comments-URI: TBD  
+Status: Draft  
+Type: Standards Track  
+Created: <date created on, in ISO 8601 (yyyy-mm-dd) format>  
+License: CC-BY-4.0  
+License-Code: Apache-2.0  
+Post-History: <dates of postings to Cardano Dev Forum, or link to thread>  
+---
+
+## Abstract
+
+this document describes the parts of the current serialization algorithm that can be improved and provides the specification and documentation needed in order to implement an optimized version of this one.
+
+## Motivation
+
+all Untyped Plutus Core Programs do need to be serialized in order to be included in a transaction (therefore stored on-chain);
+
+the current serialization algorithm is based on the [flat](http://quid2.org/docs/Flat.pdf) algorithm;
+
+this allows for a bit-oriented serialized program; but there are aspects that can be improved to further reduce the size of the serialized program;
+
+the major benefit of a uplc-optimized algorithm is the possibility to store more complex logic (or more in general, more complex transactions) on chain without the need to increase the transaction size limit.
+
+## Specification
+
+### Table of contents
+
+- [```pad0``` should add nothing](#pad0)
+- [constant's types list should work as Integer ones](#constT)
+- [remove the "type application" tag for constant types](#tyApp)
+- [```ByteString```s encoded as unsinged integer followed by the bytes](#ByteStrings)
+- [```data``` serialization
+](#data)
+
+<a name="pad0"></a>
+
+### ```pad0``` should add nothing rather than an entire byte
+
+sometimes it is necessary to allign the serialized program (partial or total) to fit into a byte;
+
+in order to do so, some padding is added based on the current bit-length of the program;
+
+the needed padding can be obtained as follows
+```ts
+pad( 8 - (currBitLength % 8) )
+```
+where the ```pad``` function could be defined as follows in typescript
+```ts
+pad( missingBits: 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 ): string
+{
+    switch( missingBits )
+    {
+        case 0: return "00000001";
+        case 1: return "1";
+        case 2: return "01";
+        case 3: return "001";
+        case 4: return "0001";
+        case 5: return "00001";
+        case 6: return "000001";
+        case 7: return "0000001";
+    }
+}
```

Comment:
What language is this? Why not describe this in Haskell?

Reply:
typescript just because I feel the description is more intuitive using it

it translates to haskell as
```hs
pad :: Int -> String
pad 0 = "00000001"
pad 1 = "1"
pad 2 = "01"
pad 3 = "001"
pad 4 = "0001"
pad 5 = "00001"
pad 6 = "000001"
pad 7 = "0000001"
pad _ = undefined
```